### ISSUE TITLE
Qwen3-VL-2B-Instruct: Forward time is much slower than other models (2.5s vs 0.01s)

### ISSUE THREAD

**Issue by Caesar107** [QwenLM/Qwen3-VL#1811]
Hi team & community,

We’ve been benchmarking **Qwen3-VL-2B-Instruct** using the `transformers` API and noticed that the **forward() pass is significantly slower** than other similar vision-language models.

###  Setup

* **Task**: Forward pass to extract hidden states (no generation).
* **Code**:

  ```python
  outputs = model(**inputs, output_hidden_states=True, return_dict=True)
  ```
* **Input**:

  * Two images
  * Size: **256x256**
  

  * GPU: GeForce RTX 5090 Graphics Cards

### Results

| Model                       | Forward Time per call | TTFT (generate) |
| --------------------------- | --------------------- | --------------- |
| Qwen3-VL-2B-Instruct        | **~2.5 seconds**      | ~4.5 seconds          |
| Eagle-3B  | ~0.01 seconds         | -               |
| Paligemma2-3b-pt-224 | ~0.04 seconds         | -               |

Note: These times are measured by repeatedly calling `model(**inputs)` in `torch.inference_mode()` and averaging over 10 runs.

---

### Question

* Is this high forward latency **expected** for Qwen3-VL models?

We suspect the vision backbone or rotary embeddings may be slow. (speed will be faster when we remove image input ~0.15s/per forward)

Would love to hear your insights or see if others are experiencing similar performance gaps. Thanks in advance!

**Comment by JJJYmmm** [comment]
Could you check the visual token numbers or the prefill seqlens across different models?

**Comment by Caesar107** [comment]
@JJJYmmm 
We compared the **visual token counts** and **prefill sequence lengths** across the three models（using same image input）:

### Visual Token Counts

| Model                  | Visual Tokens                      | Fixed? | Notes                                                                |
| ---------------------- | ---------------------------------- | ------ | -------------------------------------------------------------------- |
| **Eagle 3B** | *Depends on visual encoder config* | ❓ No   | Visual token count varies based on the encoder and image resolution. |
| **PaliGemma 3B**          | 256                                | ✅ Yes  | Always generates 256 visual tokens per image.                        |
| **Qwen3-VL 2B**           | 256                                | ✅ Yes  | Image is split into a 16×16 patch grid, totaling 256 tokens.         |

---

###  Prefill Sequence Lengths (Text Tokens)

| Model         | Text Token Count | Prefill SeqLen |
| ------------- | ---------------- | -------------- |
| **Eagle**     | 286              | 286            |
| **PaliGemma** | 258              | 258            |
| **Qwen3-VL**  | 75               | 75             |

---

### Processing Time (for reference)

| Model         | Time Taken |
| ------------- | ---------- |
| **Eagle**     | 0.0492s    |
| **PaliGemma** | 0.1475s    |
| **Qwen3-VL**  | 1.3767s    |

**Comment by Caesar107** [comment]
@JJJYmmm btw, If we remove the image input, Qwen3-VL's processing speed can reach as fast as 0.01 seconds

**Comment by JJJYmmm** [comment]
@Caesar107 that's normal because the visual tokens are much longer than text. Did you do gpu warmup before testing？

**Comment by Caesar107** [comment]
@JJJYmmm Yeah, we did warm up before testing.
We ran 10 iterations and reported the average forward（continuous）.

We also compared results with and without GPU warmup, and with/without Flash Attention enabled —
surprisingly, the difference was minimal in our case.

**Comment by JJJYmmm** [comment]
Thanks for the feedback, i would take a closer look tonight.🤗

**Comment by JJJYmmm** [comment]
My test result is normal. Maybe some wrong with your test script?

My result:

```bash
Qwen3VL-2B input length:  272
Qwen3VL-2B forward meantime: 1.630133 seconds
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████| 2/2 [00:03<00:00,  1.63s/it]
Using a slow image processor as `use_fast` is unset and a slow processor was saved with this model. `use_fast=True` will be the default behavior in v4.52, even if the model was saved with a slow processor. This will result in minor differences in outputs. You'll still be able to use a slow processor with `use_fast=False`.
Paligemma2-3B-PT-224 input length:  270
Paligemma2-3B-PT-224 forward meantime: 1.951432 seconds
```

My script:

```python
from transformers import AutoModelForImageTextToText, AutoProcessor, Qwen3VLForConditionalGeneration
from PIL import Image
import time
import torch
import gc

qwen3vl_2b_ckpt_path = "qwen3vl-2b-instruct"

paligemma2_3b_pt_224_ckpt_path = "paligemma2-3b-pt-224"

image = Image.open("cakes.png")
# resize to 256x256
image_256 = image.resize((256, 256))

image_512 = image.resize((512, 512))


# test the forward time of qwen3vl2b and paligemma2-3b-pt-224, just call model.forward

messages = [
    {
        "role": "user",
        "content": [
            {
                "type": "image",
                "image": image_512,
            },
            {"type": "text", "text": "Detect all the cakes. "},
        ],
    }
]


##### qwen3vl

model = AutoModelForImageTextToText.from_pretrained(qwen3vl_2b_ckpt_path)
processor = AutoProcessor.from_pretrained(qwen3vl_2b_ckpt_path)

inputs = processor.apply_chat_template(
    messages,
    tokenize=True,
    add_generation_prompt=True,
    return_dict=True,
    return_tensors="pt",
)
inputs = inputs.to(model.device)

print("Qwen3VL-2B input length: ", inputs.input_ids.shape[1])


# warmup 10 times
for i in range(10):
    model.forward(**inputs, output_hidden_states=True)

# test the forward time
start_time = time.perf_counter()
for i in range(10):
    model.forward(**inputs, output_hidden_states=True)
end_time = time.perf_counter()
print(f"Qwen3VL-2B forward meantime: {(end_time - start_time) / 10:.6f} seconds")

del model, inputs

# clear cuda cache and gc
torch.cuda.empty_cache()
gc.collect()

##### paligemma2-3b-pt-224
model = AutoModelForImageTextToText.from_pretrained(paligemma2_3b_pt_224_ckpt_path)
processor = AutoProcessor.from_pretrained(paligemma2_3b_pt_224_ckpt_path)

# paligemma2-3b-pt-224 doesn't have a chat template
prompt = f"<image> Detect all the cakes one by one in JSON format. "
inputs = processor(text=prompt, images=image_256, return_tensors="pt").to(model.dtype).to(model.device)

print("Paligemma2-3B-PT-224 input length: ", inputs.input_ids.shape[1])

# warmup 10 times
for i in range(10):
    model.forward(**inputs, output_hidden_states=True)

# test the forward time
start_time = time.perf_counter()
for i in range(10):
    model.forward(**inputs, output_hidden_states=True)
end_time = time.perf_counter()
print(f"Paligemma2-3B-PT-224 forward meantime: {(end_time - start_time) / 10:.6f} seconds")
```

**Comment by JJJYmmm** [comment]
qwen3vl has a downsample rate of 32 for image hight/width (16 for gemma), so I use 512x512 for qwen3vl input and 256x256 for gemma

**Comment by Caesar107** [comment]
@JJJYmmm Thanks a lot for your help — we’ll try a few more tests on our side and see how it goes.

**Comment by Caesar107** [comment]
@JJJYmmm We ran a new round of tests today, and here are our latest results:

```
Qwen3VL-2B input length: 271
Qwen3VL-2B forward meantime: 0.066998 seconds
```

On our **Windows machine with an RTX 5080**, the speed is extremely fast and looks completely normal.

However, on several of our **Linux servers**, the forward pass is consistently around **1 second or more**, even after warm-up and using the same code and image input size. Since the Windows results are much faster, we believe that ~1s on Linux is probably **not the expected performance**.

We suspect that something in our Linux environments (driver / CUDA / Transformers / image processor behavior / kernel fallback) might be causing the slowdown, and we will keep investigating to find out what is happening.

If you happen to discover anything on your side regarding Linux performance differences, please feel free to let us know as well — we’d really appreciate it.😊

**Comment by Caesar107** [comment]
@JJJYmmm We’ve finally identified the root cause of the performance issue.

When running Qwen3-VL with the **latest PyTorch 2.9.1 + cu130**,
the forward pass becomes extremely slow (around **1.3s per forward** on our side).

However, when we switch to:

```
torch==2.8.0+cu128
```

the model speeds up dramatically —
each forward pass is only **~0.06 seconds**, which matches the expected performance.



Could you please check whether you see the same behavior on your side?

If you have any idea why **PyTorch 2.9.x + cu130** causes such a slowdown for Qwen3-VL inference, we would really appreciate your insights.

**Comment by bogoconic1** [comment]
I am facing the same issue with Qwen3-VL-4B as well

```
import torch
import time
from transformers import AutoProcessor, Qwen3VLForConditionalGeneration

processor = AutoProcessor.from_pretrained('Qwen/Qwen3-VL-4B-Instruct')
image_path = '/kaggle/input/csiro-biomass/train/ID1011485656.jpg'
messages = [{'role': 'user', 'content': [{'type': 'image', 'image': image_path}, {'type': 'text', 'text': 'test'}]}]
inputs = processor.apply_chat_template(messages, tokenize=True, add_generation_prompt=True, return_dict=True, return_tensors='pt').to('cuda:0')

print('Testing WITHOUT quantization (FP16 only)...')
start = time.time()
model = Qwen3VLForConditionalGeneration.from_pretrained(
  'Qwen/Qwen3-VL-4B-Instruct',
  torch_dtype=torch.float16,
  device_map='cuda:0'
)
print(f'Model load: {time.time()-start:.2f}s')

start = time.time()
with torch.no_grad():
  image_embeds, _ = model.get_image_features(pixel_values=inputs['pixel_values'], image_grid_thw=inputs['image_grid_thw'])
torch.cuda.synchronize()
print(f'get_image_features (FP16): {time.time()-start:.4f}s') # 0.67s with 2.8.0, 47s with 2.9.1 
```

**Comment by bogoconic1** [comment]
It is caused by a known PyTorch issue [here](https://github.com/pytorch/pytorch/issues/166122) in 2.9.x where ```Conv3d``` latency has significant regressions due to cuDNN being disabled. They are actively fixing it.

Here are some logs of the `get_image_features` that show the breakdown of the `47s with 2.9.1`

<img width="2714" height="1192" alt="Image" src="https://github.com/user-attachments/assets/4e9f0994-c4d3-48f3-8191-d43a58720802" />

<img width="2478" height="394" alt="Image" src="https://github.com/user-attachments/assets/8722c7ff-6a54-4865-90da-ca56be0984c6" />

so downgrading `torch` to `2.8.0` is the correct solution for now

**Comment by Caesar107** [comment]
@bogoconic1 Thanks a lot for the clarification — that really explains everything!

**Comment by weathon** [comment]
for me, on A6000, batch_size=8 for Qwen 3 Embedding, one forward pass takes about 65s, that is insane. After downgraded to '2.8.0+cu128', it still takes about 1.6s/it with BS=8
